File: backend/schemas/analytics.py
# schemas/analytics.py
"""
Analytics schemas for dashboard metrics and reporting.
"""

from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel

# ============== Dashboard Overview ==============


class DashboardOverview(BaseModel):
    """Main dashboard summary metrics."""

    # Counts
    total_products: int
    products_with_auto_pricing: int
    total_competitors: int

    # Alerts
    unread_alerts: int
    alerts_today: int

    # Recommendations
    pending_recommendations: int
    applied_recommendations_7d: int

    # Sentiment (last 24h)
    average_sentiment: float | None = None
    sentiment_trend: str = "stable"  # "improving", "declining", "stable"
    total_mentions_24h: int = 0


class ProductSummary(BaseModel):
    """Product card for dashboard."""

    id: str
    name: str
    sku: str | None = None
    current_price: Decimal
    base_price: Decimal
    price_change_percent: float
    sentiment_score: float | None = None
    mention_count_24h: int = 0
    has_pending_recommendation: bool = False
    auto_pricing_enabled: bool = False


# ============== Recommendation Stats ==============


class RecommendationStats(BaseModel):
    """Recommendation performance metrics."""

    total_generated: int
    total_applied: int
    total_rejected: int
    total_expired: int
    total_pending: int

    approval_rate: float  # applied / (applied + rejected)
    avg_confidence: float | None = None
    avg_price_change_percent: float | None = None


# ============== Sentiment Analytics ==============


class SentimentDataPoint(BaseModel):
    """Single point in sentiment timeline."""

    timestamp: datetime
    score: float
    mention_count: int


class SentimentAnalytics(BaseModel):
    """Sentiment trends over time."""

    product_id: str | None = None
    period_days: int
    current_score: float | None = None
    previous_score: float | None = None
    change: float | None = None
    trend: str = "stable"
    timeline: list[SentimentDataPoint] = []


# ============== Revenue Impact ==============


class RevenueImpact(BaseModel):
    """Revenue impact from price changes."""

    period_days: int
    total_price_changes: int
    avg_change_percent: float
    products_increased: int
    products_decreased: int

    # If outcome tracking exists
    estimated_revenue_impact: Decimal | None = None


# ============== Alert Analytics ==============


class AlertAnalytics(BaseModel):
    """Alert statistics."""

    total_alerts_7d: int
    by_type: dict  # {"sentiment_change": 5, "price_recommendation": 12, ...}
    by_severity: dict  # {"low": 10, "medium": 5, "high": 2, "critical": 1}
    avg_resolution_time_hours: float | None = None

File: frontend/types/analytics.ts
// Analytics domain types
// AUTO-SYNCED with backend via openapi-typescript
// Last synced: 2026-01-08
// Source: components["schemas"]["DashboardOverview"], RecommendationStats, etc.

import type { SentimentDataPoint, SentimentTrendDirection } from './sentiment';
import type { ProductSummary } from './product';

// AP-013/sprint-1: RulePerformanceSummary was defined here AND in outcome.ts,
// causing a duplicate export collision when both are re-exported from types/index.ts.
// Removed local definition — import from outcome.ts as the single source of truth.
// AccuracyStatsResponse uses RulePerformanceSummary, so we import it here for use below.
import type { RulePerformanceSummary } from './outcome';
export type { RulePerformanceSummary };

// ============================================
// DASHBOARD OVERVIEW
// ============================================

/**
 * Dashboard overview stats
 * Matches: components["schemas"]["DashboardOverview"]
 */
export interface DashboardOverview {
  total_products: number;
  products_with_auto_pricing: number;
  total_competitors: number;
  unread_alerts: number;
  alerts_today: number;
  pending_recommendations: number;
  applied_recommendations_7d: number;
  average_sentiment: number | null;
  sentiment_trend: 'improving' | 'declining' | 'stable';
  total_mentions_24h: number;
}

// ============================================
// RECOMMENDATION STATS
// ============================================

/**
 * Recommendation statistics
 * Matches: components["schemas"]["RecommendationStats"]
 */
export interface RecommendationStats {
  total_generated: number;
  total_applied: number;
  total_rejected: number;
  total_expired: number;
  total_pending: number;
  approval_rate: number;
  avg_confidence: number | null;
  avg_price_change_percent: number | null;
}

// ============================================
// ALERT ANALYTICS
// ============================================

/**
 * Alert analytics for dashboard
 * Matches: components["schemas"]["AlertAnalytics"]
 */
export interface AlertAnalytics {
  total_alerts_7d: number;
  by_type: Record<string, unknown>;
  by_severity: Record<string, unknown>;
  avg_resolution_time_hours: number | null;
  by_status?: Record<string, number>;
}

// ============================================
// SENTIMENT ANALYTICS
// ============================================

/**
 * Sentiment trend with analytics context
 */
export interface SentimentTrendAnalytics {
  product_id: string | null;
  period_days: number;
  current_score: number | null;
  previous_score: number | null;
  change: number | null;
  trend: SentimentTrendDirection;
  timeline: SentimentDataPoint[];
}

// ============================================
// ACCURACY STATS
// ============================================

/**
 * Per-rule-type accuracy breakdown
 */
export interface RuleTypeStats {
  count: number;
  positive: number;
  revenue_impact: number;
  success_rate: number;
}

/**
 * Accuracy statistics for recommendations
 * Matches: components["schemas"]["AccuracyStatsResponse"]
 */
export interface AccuracyStatsResponse {
  period_days: number;
  total_outcomes: number;
  positive_count: number;
  negative_count: number;
  neutral_count: number;
  inconclusive_count: number;
  overall_success_rate: string;
  avg_outcome_score: string;
  total_revenue_impact: string;
  avg_revenue_change_percent: string | null;
  by_rule_type: Record<string, RuleTypeStats>;
  top_performing_rules: RulePerformanceSummary[];
  worst_performing_rules: RulePerformanceSummary[];
}

// Re-export ProductSummary for convenience
export type { ProductSummary };



